### PR TITLE
[BEAM-564] Updates sources to report consumed and remaining number of split points

### PR DESCRIPTION
**Pull request by chamikaramj**
Adds several methods to the RangeTracker interface to support this. Please see comments for details.

Updates AvroSource and LineSource (test) to report split points properly.

Runners can use this information to determine the amount of remaining and consumed parallelism of source read operations. Java SDK sources framework already supports reporting these signals.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @aaltay Ahmet, Could you do an overall review of this PR ?

R: @dhalperi Dan, Could you review this focusing on logic for generating consumed and remaining number of split points (since you wrote the corresponding Java PR).

Thanks.

File: sdks/python/apache_beam/io/avroio.py [hunk 1]
```diff
@@ -137,6 +139,26 @@ def offset(self):
...
 
   def read_records(self, file_name, range_tracker):
+    current_block_offset = -1
+    current_block_size = -1
+
+    def _callback(stop_position):
+      if range_tracker.done():
+        return 0
+
+      # Locking progressing to next block  to provide a consistent result
```
Comment (aaltay) [inline sdks/python/apache_beam/io/avroio.py]: Could you clarify the `progressing` part in the comment. What are we preventing here?

File: sdks/python/apache_beam/io/avroio.py [hunk 2]
```diff
@@ -137,6 +139,26 @@ def offset(self):
...
+      with self._progress_lock:
+        if (current_block_offset + current_block_size >=
+            range_tracker.stop_position()):
+          # This block ends after the last byte that belongs to the position
+          # range of the source. Hence this has to be the last block that
+          # belongs to the position range.
+          return 1
+        return filebasedsource.FileBasedSource.remaining_split_points_helper(
```
Comment (aaltay) [inline sdks/python/apache_beam/io/avroio.py]: Would this return anything other than `iobase.RangeTracker.SPLIT_POINTS_UNKNOWN`. It seems like the first two conditions are already checked here.

File: sdks/python/apache_beam/io/filebasedsource.py [hunk 3]
```diff
@@ -179,6 +180,58 @@ def read_records(self, file_name, offset_range_tracker):
   def splittable(self):
     return self._splittable
 
+  @staticmethod
+  def remaining_split_points_helper(
+      stop_position, current_position, done_reading):
+    """A util function for implementing a remaining splits points callback.
```
Comment (aaltay) [inline sdks/python/apache_beam/io/filebasedsource.py]: _util function_ -> _utility function_ ?

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 4]
```diff
@@ -285,6 +294,19 @@ def test_dataflow_pattern_unsplittable(self):
     assert len(expected_data) == 200
     self._run_dataflow_test(pattern, expected_data, False)
 
+  def test_remaining_split_points_helper(self):
```
Comment (aaltay) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Could you add another case for `SPLIT_POINTS_UNKNOWN` branch.

File: sdks/python/apache_beam/io/fileio.py [hunk 5]
```diff
@@ -796,17 +796,24 @@ def __iter__(self):
         # that the end offset of the range may be smaller than the original
         # end offset defined when creating the reader due to reader accepting
         # a dynamic split request from the service.
-        return
+        break
```
Comment (aaltay) [inline sdks/python/apache_beam/io/fileio.py]: This will yield one more line as a result of this change after `try_claim` fails, is this intentional?

File: sdks/python/apache_beam/io/iobase.py [hunk 6]
```diff
@@ -659,6 +675,159 @@ def fraction_consumed(self):
...
+    Args:
+      callback: a function that takes the a single parameter, a stop position,
+                and returns remaining number of split points for the source read
+                operation that is calling this function. Value returned from
+                callback should be either a integer larger than or equal to zero
+                or ``RangeTracker.SPLIT_POINTS_UNKNOWN``.
+    """
+    pass
+
+  def set_done(self):
```
Comment (aaltay) [inline sdks/python/apache_beam/io/iobase.py]: You could use `@property` for `done` and `set_done` and only have a `done` property.

File: sdks/python/apache_beam/io/range_trackers.py [hunk 7]
```diff
@@ -101,11 +106,40 @@ def _validate_record_start(self, record_start, split_point):
...
+      if self._split_points_remaining_callback:
+        if (self._split_points_remaining ==
+            iobase.RangeTracker.SPLIT_POINTS_UNKNOWN):
+          if not self._first_claim_done:
+            # Trying to set the initial value for remaining number of split
+            # points.
+            # We only try to do this at first claim and when stop position gets
+            # updated so that we do not unnecessarily keep calling the callback
+            # for every call to try_claim().
+            self._split_points_remaining = (
```
Comment (aaltay) [inline sdks/python/apache_beam/io/range_trackers.py]: Don't you need to check for the same condition `self._split_points_remaining <= 0`

File: sdks/python/apache_beam/io/range_trackers.py [hunk 8]
```diff
@@ -101,11 +106,40 @@ def _validate_record_start(self, record_start, split_point):
...
+                self._split_points_remaining_callback(self.stop_position()))
+        else:
+          # Split points remaining already set. Updating.
+          self._split_points_remaining -= 1
+          if self._split_points_remaining <= 0:
+            raise ValueError('try_claim() for position %r was successful but '
```
Comment (aaltay) [inline sdks/python/apache_beam/io/range_trackers.py]: This messages is not correct from a user's perspective. Call to the `try_claim` is **unsuccessful** in this case.

File: sdks/python/apache_beam/io/range_trackers.py [hunk 9]
```diff
@@ -285,6 +354,17 @@ def fraction_consumed(self):
                        ' consumed due to positions being opaque strings'
                        ' that are interpreted by the service')
 
+  def split_points(self):
+    with self._lock:
```
Comment (aaltay) [inline sdks/python/apache_beam/io/range_trackers.py]: This is for my information. What is the need for locks in this files? Is it being accessed by multiple threads?

File: sdks/python/apache_beam/io/avroio.py (hunk 2 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/avroio.py]: This will return 1 if current block starts at the last byte of the range (previousl we only checked the next block).

File: sdks/python/apache_beam/io/filebasedsource_test.py (hunk 4 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Done.

File: sdks/python/apache_beam/io/range_trackers.py (hunk 7 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/range_trackers.py]: Done.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/fileio.py]: Since we break the while loop, we simpl invoke set_done() and return. Don't think this yields an extra result.

File: sdks/python/apache_beam/io/iobase.py (hunk 6 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/iobase.py]: I thout about that but seems like it's hard (and error prone) to override properties in sub-classes. We expect some users to override these methods (for example, OffsetRangeTracker).

File: sdks/python/apache_beam/io/range_trackers.py (hunk 9 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/range_trackers.py]: Depending on the runner, this may be invoked by multiple threads.

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/avroio.py]: We want to make sure that parameters within the callback are for a single block. Updated the comment.

File: sdks/python/apache_beam/io/filebasedsource.py (hunk 3 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/filebasedsource.py]: Done.

File: sdks/python/apache_beam/io/range_trackers.py (hunk 8 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/range_trackers.py]: Done.

**Comment by chamikaramj** [comment]
Thanks. PTAL.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (aaltay) [inline sdks/python/apache_beam/io/fileio.py]: You are right, it would not lead an extra yield. I misread the indentation of the yield, my mistake.

**Comment by aaltay** [comment]
LGTM. Thank you.

**Comment by chamikaramj** [comment]
Thanks Ahmet.

**Comment by chamikaramj** [comment]
R: @dhalperi

**Comment by chamikaramj** [comment]
Ping :)

**Comment by chamikaramj** [comment]
R: @lukecwik Luke, could you merge since Dan is busy. Note that this was already reviewed by Ahmet. But I'll be happy to address any further comments.

Thanks.

File: sdks/python/apache_beam/io/avroio.py [hunk 10]
```diff
@@ -94,6 +95,7 @@ def __init__(self, file_pattern, min_bundle_size=0):
     self._avro_schema = None
     self._codec = None
     self._sync_marker = None
+    self._progress_lock = threading.Lock()
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: does Python have any annotations you would use like Java's `@GuardedBy` to enable tools to identify locking bugs?

File: sdks/python/apache_beam/io/avroio.py [hunk 11]
```diff
@@ -137,6 +139,28 @@ def offset(self):
...
 
   def read_records(self, file_name, range_tracker):
+    current_block_offset = -1
+    current_block_size = -1
+
+    def _callback(stop_position):
+      if range_tracker.done():
+        return 0
+
+      # Locking progressing to next block  to provide a consistent result.
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: double `  `

File: sdks/python/apache_beam/io/avroio.py (hunk 10 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: suggest you at least comment which state this lock guards

File: sdks/python/apache_beam/io/avroio.py [hunk 12]
```diff
@@ -137,6 +139,28 @@ def offset(self):
       return self._offset
 
   def read_records(self, file_name, range_tracker):
+    current_block_offset = -1
+    current_block_size = -1
+
+    def _callback(stop_position):
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: is there a better name for `_callback`? Maybe `_split_points_remaining` ?

Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: and/or should this function have documentation as to what it returns?

File: sdks/python/apache_beam/io/avroio.py [hunk 13]
```diff
@@ -156,12 +180,17 @@ def read_records(self, file_name, range_tracker):
           return
         next_block = self.read_next_block(f)
         if next_block:
+          with self._progress_lock:
+            current_block_offset = next_block.offset()
+            current_block_size = next_block.size()
+
           for record in next_block.records():
             yield record
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: how do you update progress within blocks?

File: sdks/python/apache_beam/io/avroio_test.py [hunk 14]
```diff
@@ -145,6 +146,29 @@ def test_read_with_splitting_pattern(self):
     expected_result = self.RECORDS * 3
     self._run_avro_test(pattern, 100, True, expected_result)
 
+  def test_read_split_points(self):
+    file_name = self._write_data(count=20)
+    unsplit_source = AvroSource(file_name)
+
+    # desired_bundle_size should be large enough to make sure that first split
+    # is not a trivial one.
+    splits = [split for split in unsplit_source.split(1000)]
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio_test.py]: isn't this just `list(unsplit_source.split(1000))` ?

Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio_test.py]: e.g.,

```python
>>> list(xrange(10))
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> xrange(10)
xrange(10)
```

File: sdks/python/apache_beam/io/avroio_test.py [hunk 15]
```diff
@@ -145,6 +146,29 @@ def test_read_with_splitting_pattern(self):
...
+    source = splits[0].source
+    start_position = splits[0].start_position
+    stop_position = splits[0].stop_position
+
+    range_tracker = source.get_range_tracker(start_position, stop_position)
+    assert isinstance(range_tracker, iobase.RangeTracker)
+    reader = source.read(range_tracker)
+    for _ in reader:
+      self.assertEquals((0, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN),
+                        range_tracker.split_points())
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio_test.py]: how do you guarantee the last block has exactly one record? I would think you'd eventually get `(0, 1)` in real Avro files.

File: sdks/python/apache_beam/io/filebasedsource.py [hunk 16]
```diff
@@ -213,6 +213,58 @@ def read_records(self, file_name, offset_range_tracker):
   def splittable(self):
     return self._splittable
 
+  @staticmethod
+  def remaining_split_points_helper(
+      stop_position, current_position, done_reading):
+    """A utility function for implementing a remaining splits points callback.
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/filebasedsource.py]: "remaining split points" or "split points remaining"?

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 17]
```diff
@@ -348,6 +357,25 @@ def test_read_zlib_pattern(self):
...
+    # Current record is after stop position but we haven't reached the next
+    # split point
+    self.assertEqual(
+        1, FileBasedSource.remaining_split_points_helper(100, 110, False))
+    # After RangeTracker.set_done() is invoked, there should not be any
+    # remaining split points.
+    self.assertEqual(
+        0, FileBasedSource.remaining_split_points_helper(100, 110, True))
+
+    # If we are at a position < (stop_position -1), FileBasedSource cannot
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: either `stop_position - 1` or `stop_position-1`, but be consistent on spaces or I think `-1` is a number here.

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 18]
```diff
@@ -41,6 +41,14 @@
 class LineSource(FileBasedSource):
 
   def read_records(self, file_name, range_tracker):
+    current = -1
+
+    def _callback(stop_position):
+      return FileBasedSource.remaining_split_points_helper(
+          stop_position, current, range_tracker.done())
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: why do you not need a progress lock here? there are two calls made to things that may be changed by the read loop: `current` and `range_tracker`.

File: sdks/python/apache_beam/io/fileio.py [hunk 19]
```diff
@@ -796,17 +796,24 @@ def __iter__(self):
         # that the end offset of the range may be smaller than the original
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/fileio.py]: might be nice to stay stylistically consistent unless there is some reason I'm missing.

```python
try:
   ...
finally:
  ...
```
then break vs return is not a big deal.

File: sdks/python/apache_beam/io/iobase.py [hunk 20]
```diff
@@ -659,6 +675,159 @@ def fraction_consumed(self):
...
+    By default ``RangeTracker` returns ``RangeTracker.SPLIT_POINTS_UNKNOWN`` for
+    both consumed and remaining number of split points, which indicates that the
+    number of split points consumed and remaining is unknown.
+
+    Returns:
+      a pair that gives consumed and remaining number of split points. Each
+      value should be an integer that is larger than or equal to zero or
+      ``RangeTracker.SPLIT_POINTS_UNKNOWN``.
+    """
+    return (RangeTracker.SPLIT_POINTS_UNKNOWN,
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/iobase.py]: should a failed `try_claim` cause `done` to be set? why or why not?

Comment (dhalperi) [inline sdks/python/apache_beam/io/iobase.py]: should you guard against other methods being called once `done` is set, e.g., no `try_claim` after done?

File: sdks/python/apache_beam/io/range_trackers.py [hunk 21]
```diff
@@ -101,11 +106,45 @@ def _validate_record_start(self, record_start, split_point):
...
       self._last_record_start = record_start
+      self._split_points_seen += 1
+      if self._split_points_remaining_callback:
+        if (self._split_points_remaining ==
+            iobase.RangeTracker.SPLIT_POINTS_UNKNOWN):
+          if not self._first_claim_done:
+            # Trying to set the initial value for remaining number of split
+            # points.
+            # We only try to do this at first claim and when stop position gets
+            # updated so that we do not unnecessarily keep calling the callback
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/range_trackers.py]: why is this unnecessary -- advancing the position can actually change the number of split points remaining, right?

File: sdks/python/apache_beam/io/range_trackers_test.py [hunk 22]
```diff
@@ -157,6 +158,28 @@ def test_try_return_record_non_monotonic(self):
...
+                     (0, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
+    self.assertTrue(tracker.try_claim(140))
+    self.assertEqual(tracker.split_points(),
+                     (1, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
+    self.assertTrue(tracker.try_claim(190))
+    self.assertEqual(tracker.split_points(),
+                     (2, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
+    self.assertTrue(tracker.try_claim(199))
+    self.assertEqual(tracker.split_points(),
+                     (3, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/range_trackers_test.py]: why is this not `(3, 1)`?

File: sdks/python/apache_beam/io/range_trackers_test.py [hunk 23]
```diff
@@ -318,6 +341,31 @@ def test_try_test_split_at_position_finite_range(self):
...
+                     (1, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
+    self.assertTrue(tracker.try_claim(self.bytes_to_position([3, 6, 8])))
+    self.assertEqual(tracker.split_points(),
+                     (2, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
+    self.assertTrue(tracker.try_claim(self.bytes_to_position([4, 255, 255])))
+    self.assertEqual(tracker.split_points(),
+                     (3, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
+    self.assertFalse(tracker.try_claim(self.bytes_to_position([5, 1, 0])))
+    self.assertEqual(tracker.split_points(),
+                     (3, iobase.RangeTracker.SPLIT_POINTS_UNKNOWN))
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/range_trackers_test.py]: why isn't this `(4, UNKNOWN)`?

File: sdks/python/apache_beam/io/avroio.py [hunk 24]
```diff
@@ -19,6 +19,7 @@
 
 import os
 import StringIO
```
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: `cStringIO as StringIO`?

**Comment by chamikaramj** [comment]
R: @robertwb

File: sdks/python/apache_beam/io/avroio.py (hunk 11 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio.py]: No, I meant there is a double ` ` character between `"block  to"` ;)

File: sdks/python/apache_beam/io/avroio_test.py (hunk 15 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/avroio_test.py]: The entire point of remaining parallelism is that once you start reading the last Avro block you can tell it's the last Avro block. In other words, remaining parallelism should go `UNKNOWN`, then `1`, then `0`. If we only go `UNKNOWN` then `0` then remaining parallelism is useless.

File: sdks/python/apache_beam/io/iobase.py (hunk 20 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/iobase.py]: @jkff -- please chime in. In java, and I think for SDF, it is important that a failed `try_claim` declares that the reader is done.

File: sdks/python/apache_beam/io/range_trackers.py (hunk 21 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/range_trackers.py]: What happens if the number of split points remaining is unknown until midway through reading -- after start, but before done.

This seems overly complex logic with poor results.

File: sdks/python/apache_beam/io/range_trackers_test.py (hunk 23 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/range_trackers_test.py]: Consumed should become 4 after failed `try_claim`, and I think remaining should become `0` after that (and definitely `0` after `set_done`).

File: sdks/python/apache_beam/io/filebasedsource_test.py (hunk 18 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: The range tracker lock is a different lock than the source lock. You need them both.

File: sdks/python/apache_beam/io/iobase.py (hunk 20 above)
Comment (jkff) [inline sdks/python/apache_beam/io/iobase.py]: Correct, a failed try_claim must end the bundle. It is the case in all current incarnations of this API - Java readers, SDF, C++ readers. Please do not introduce inconsistency.

File: sdks/python/apache_beam/io/iobase.py [hunk 25]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+
+    ** Determining consumed and remaining number of split points **
+
+    Consider the following examples: (1) An input that can be read in parallel
+    down to the individual records, is called "perfectly splittable".
+    (2) a "block-compressed" file format such as ``avroio``, in which a block of
+    records has to be read as a whole, but different blocks can be read in
+    parallel. (3) an "unsplittable" input such as a cursor in a database.
+
+    consumed number of split points:
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: Capitalize? Or are you trying to enumerate some examples here, in which case I would rephrase this.

File: sdks/python/apache_beam/io/iobase.py [hunk 26]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+        This condition holds independent of whether the input is splittable.
+      * For a non-empty, finished source read (in which
+        ``RangeTracker.try_claim()`` was invoked at least once and
+        ``RangeTracker.set_done()`` has been invoked), the value returned must
+        be at least 1 and should equal the total number of split points in the
+        source.
+      * For example (1): After returning record #30 (starting at 1) out of 50 in
+        a perfectly splittable 50-record input, this value should be 29. When
+        finished, the number of split points consumed should be 50.
+      * For example (2): In a block-compressed value consisting of 5 blocks, the
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: This isn't correct. It stays at 0 until the first record of the second block is *claimed*. Remember, the range tracker never sees what gets returned.

File: sdks/python/apache_beam/io/iobase.py [hunk 27]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+    remaining is known:
+
+    * Any source read for which ``RangeTracker.set_done()`` has not been
+      invoked, remaining number of split points should not be zero, because this
+      unfinished read itself represents a unfinished split point. This
+      condition holds independent of whether the input is splittable.
+    * A finished read, for which RangeTracker.set_done() has been invoked,
+      remaining number of split points is 0. This condition holds independent of
+      whether the input is splittable.
+    * For example 1: After returning record #30 (starting at 1) out of 50 in a
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: example (1). Similarly below.

File: sdks/python/apache_beam/io/iobase.py [hunk 28]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+      perfectly splittable 50-record input, remaining number of split points
+      should be 21 (20 remaining + 1 current) if the total number of records is
+      known.
+    * For example 2: After returning a record in block 3 in a block-compressed
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: I think this section needs to be reworded in terms of blocks/records claimed, not records returned, as that (plus split requests) is the key time these numbers change. 

It might also be clearer to go through the examples one at a time, e.g. discussing the lifetime (consumed and remaining) for totally splittable sources first, then block sources, then unsplittable.

File: sdks/python/apache_beam/io/iobase.py [hunk 29]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+    By invoking this function, a ``BoundedSource`` can set a callback function
+    that may get invoked by the RangeTracker to determine the number of
+    remaining (unconsumed) split points. The callback function accepts a single
+    parameter, a stop position for the BoundedSource (stop_position). If the
+    record currently being consumed by the ``BoundedSource`` is at position
+    current_position, callback should return the number of split points within
+    the range [current_position, stop_position). Note that this should include
+    the split point currently being consumed by the source.
+
+    The given callback must only be invoked after acquiring the lock of the
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: I think the instructions to the consumer of this method and the implementer of this method (range tracker) need to be called called out separately. (E.g. must vs. will.)

Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: It seems, however, that care still needs to be taken as this may be called from a different thread than the one that set the callback...

File: sdks/python/apache_beam/io/iobase.py [hunk 30]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+    """
+    return (RangeTracker.SPLIT_POINTS_UNKNOWN,
+            RangeTracker.SPLIT_POINTS_UNKNOWN)
+
+  def set_split_points_remaining_callback(self, callback):
+    """Sets a callback function for determining split points remaining.
+
+    By invoking this function, a ``BoundedSource`` can set a callback function
+    that may get invoked by the RangeTracker to determine the number of
+    remaining (unconsumed) split points. The callback function accepts a single
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: Wouldn't it be better to take two arguments, last_claimed + stop_position?

File: sdks/python/apache_beam/io/iobase.py (hunk 6 above)
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: When and why would we require a user to call this method?

File: sdks/python/apache_beam/io/iobase.py [hunk 31]
```diff
@@ -659,6 +675,167 @@ def fraction_consumed(self):
...
+    while this callback is being invoked.
+
+    Args:
+      callback: a function that takes a single parameter, a stop position,
+                and returns remaining number of split points for the source read
+                operation that is calling this function. Value returned from
+                callback should be either a integer larger than or equal to zero
+                or ``RangeTracker.SPLIT_POINTS_UNKNOWN``.
+    """
+    pass
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: Are we sure we want the default implementation to ignore this? The relationship between this and split_points() above should be called out.

File: sdks/python/apache_beam/io/range_trackers.py (hunk 21 above)
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: try_claim is called quite frequently; we should only invoke the callback when the reporter thread asks for the number of split points. Here all we should do is increment _split_points_seen.

File: sdks/python/apache_beam/io/range_trackers.py [hunk 32]
```diff
@@ -100,15 +105,51 @@ def _validate_record_start(self, record_start, split_point):
...
+            # for every call to try_claim().
+            self._split_points_remaining = (
+                self._split_points_remaining_callback(self.stop_position()))
+            if self._split_points_remaining <= 0:
+              raise ValueError(
```
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: Note, BTW, that "remaining" is an estimate. Throwing an error if we were off by one would be bad.

File: sdks/python/apache_beam/io/range_trackers.py [hunk 33]
```diff
@@ -143,6 +184,12 @@ def try_split(self, split_offset):
       split_fraction = (float(split_offset - self._start_offset) / (
           self._stop_offset - self._start_offset))
 
+      if self._split_points_remaining_callback:
```
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: Don't call this here (and elsewhere--just remove the _split_points_remaining).

File: sdks/python/apache_beam/io/range_trackers.py [hunk 34]
```diff
@@ -167,6 +214,28 @@ def position_at_fraction(self, fraction):
     return int(math.ceil(self.start_position() + fraction * (
         self.stop_position() - self.start_position())))
 
+  def split_points(self):
+    with self._lock:
+      if self._split_points_seen == 0:
```
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: Seems like this logic will be common to all sources. Lift?

File: sdks/python/apache_beam/io/range_trackers.py [hunk 35]
```diff
@@ -232,17 +302,25 @@ def _validate_decoded_group_start(self, decoded_group_start, split_point):
                              else 'non-')))
 
   def try_claim(self, decoded_group_start):
+    assert not self.done()
```
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: Here (and elsewhere). This assert is redundant with the check below.

File: sdks/python/apache_beam/io/range_trackers.py [hunk 36]
```diff
@@ -324,3 +413,15 @@ def set_current_position(self, position):
 
   def fraction_consumed(self):
     return self._range_tracker.fraction_consumed()
+
+  def split_points(self):
+    return self._range_tracker.split_points()
```
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: This isn't correct for an unsplittable source.

**Comment by chamikaramj** [comment]
Closing this for now till I do more testing on whether these signals are needed.
